### PR TITLE
Closes #2957: Dot Function

### LINKED ISSUE #2957: dot function
Add ak.dot function to match numpy:
https://numpy.org/doc/stable/reference/generated/numpy.dot.html

### PR DESCRIPTION
**Pull request by jaketrookman**
This PR (Closes #2957) adds the dot function to match Numpy functionality and testing.

### CODE REVIEW

**Review by ajpotts — COMMENTED** [review]
Looks great except for the one comment about throwing a ValueError.

File: arkouda/pdarrayclass.py [hunk 1]
```diff
@@ -2236,6 +2238,53 @@ def sum(pda: pdarray) -> np.float64:
...
+    >>> ak.dot(x,2)
+    array([4 6])
+    """
+    if (
+        not isinstance(pda1, pdarray)
+        and not isinstance(pda2, pdarray)
+        and ((pda1 is akuint64 and pda2 is akint64) or (pda1 is akint64 and pda2 is akuint64))
+    ):
+        raise TypeError(f"incompatible types {type(pda1)}, {type(pda2)}")
+    if isinstance(pda1, pdarray) and isinstance(pda2, pdarray):
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply arkouda/pdarrayclass.py:2282]: Thank you for the catch, added back in

File: arkouda/pdarrayclass.py [hunk 2]
```diff
@@ -2236,6 +2238,56 @@ def sum(pda: pdarray) -> np.float64:
     return parse_single_value(cast(str, repMsg))
 
 
+@typechecked
+def dot(
+    pda1: Union[np.int64, np.float64, np.uint64, pdarray],
+    pda2: Union[np.int64, np.float64, np.uint64, pdarray],
+) -> Union[np.int64, np.float64, np.uint64, pdarray]:
+    """
+    Returns the sum of the product of two arrays of the same size (the dot product) or
```
Comment (pierce314159) [inline arkouda/pdarrayclass.py:2247]: maybe we reword this to be:
```suggestion
    Returns the sum of the elementwise product of two arrays of the same size (the dot product) or
```
because otherwise the product of 2 arrays is a bit ambiguous

File: arkouda/pdarrayclass.py [hunk 3]
```diff
@@ -2236,6 +2238,56 @@ def sum(pda: pdarray) -> np.float64:
...
+    ----------
+    pda1 : Union[numeric_scalars, pdarray]
+
+    pda2 : Union[numeric_scalars, pdarray]
+
+
+    Returns
+    -------
+    Union[numeric_scalars, pdarray]
+        The sum of the product pda1 and pda2 or the product of a singleton element and an array.
```
Comment (pierce314159) [inline arkouda/pdarrayclass.py:2260]:
```suggestion
        The sum of the elementwise product pda1 and pda2 or the product of a singleton element and an array.
```

**Review by pierce314159 — COMMENTED** [review]
Looks good! one small note on the docstring

File: arkouda/pdarrayclass.py (hunk 2 above)
Comment (jaketrookman) [reply arkouda/pdarrayclass.py]: I agree, that is much more clear. Fixed

File: arkouda/pdarrayclass.py (hunk 3 above)
Comment (jaketrookman) [reply arkouda/pdarrayclass.py]: Fixed

**Review by pierce314159 — APPROVED** [review]
